### PR TITLE
order implementation

### PR DESCRIPTION
**Pull request by ADIGrimm**

### CODE REVIEW

File: src/main/java/online/bookstore/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+@RequiredArgsConstructor
+@RestController
+public class OrderController implements UserContextHelper {
+    private final OrderService orderService;
+
+    @Operation(summary = "Get user order history",
+            description = "Return list of all orders of user")
+    @GetMapping
+    @PreAuthorize("hasAnyRole('ROLE_USER')")
+    public Page<OrderDto> getAll(Authentication authentication, Pageable pageable) {
```
Comment (Elena-Bruyako) [inline src/main/java/online/bookstore/controller/OrderController.java]: you can add
```suggestion
    public Page<OrderDto> getAll(Authentication authentication, @ParameterObject @PageableDefault Pageable pageable) {
```

File: src/main/java/online/bookstore/controller/OrderController.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        return orderService.getAll(getUserId(authentication), pageable);
+    }
+
+    @Operation(summary = "Create order",
+            description = "Create order")
+    @PostMapping
+    @PreAuthorize("hasAnyRole('ROLE_USER')")
+    public OrderDto createOrder(Authentication authentication, @RequestBody String address) {
```
Comment (Elena-Bruyako) [inline src/main/java/online/bookstore/controller/OrderController.java]: It's better to create Dto

```suggestion
    public OrderDto createOrder(Authentication authentication, @RequestBody @Valid CreateOrderRequestDto orderDto) {
```

File: src/main/java/online/bookstore/controller/OrderController.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        return orderService.create(getUserId(authentication), address);
+    }
+
+    @Operation(summary = "Update order status", description =
+            "Updates the order status to one of these: delivered, cancelled, pending")
+    @PutMapping("/{id}")
+    @PreAuthorize("hasAnyRole('ROLE_ADMIN')")
+    public OrderDto updateStatus(@PathVariable Long orderId, @RequestBody Order.Status status) {
```
Comment (Elena-Bruyako) [inline src/main/java/online/bookstore/controller/OrderController.java]: same here

File: src/main/java/online/bookstore/controller/OrderItemController.java [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+import org.springframework.security.core.Authentication;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@RequestMapping("/orders/{orderId}/items")
+@RequiredArgsConstructor
+@RestController
+public class OrderItemController implements UserContextHelper {
```
Comment (Elena-Bruyako) [inline src/main/java/online/bookstore/controller/OrderItemController.java]: remove controller, all endpoints should be in OrderController

File: src/main/java/online/bookstore/model/Order.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+    private LocalDateTime orderDate;
+    @NotNull
+    private String shippingAddress;
+    @OneToMany(mappedBy = "order",
+            cascade = CascadeType.ALL,
+            orphanRemoval = true)
+    private Set<OrderItem> orderItems = new HashSet<>();
+
+    public enum Status {
+        Pending,
```
Comment (Elena-Bruyako) [inline src/main/java/online/bookstore/model/Order.java]: use upper case only for enum

File: src/main/java/online/bookstore/model/OrderItem.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+@Setter
+@Table(name = "order_items")
+public class OrderItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "order_id", nullable = false)
+    private Order order;
+    @OneToOne
```
Comment (Elena-Bruyako) [inline src/main/java/online/bookstore/model/OrderItem.java]: Add lazy and joincolumn

File: src/main/java/online/bookstore/service/impl/OrderItemServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+import online.bookstore.mapper.OrderItemMapper;
+import online.bookstore.repository.order.OrderItemRepository;
+import online.bookstore.service.OrderItemService;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class OrderItemServiceImpl implements OrderItemService {
```
Comment (Elena-Bruyako) [inline src/main/java/online/bookstore/service/impl/OrderItemServiceImpl.java]: remove, replace logic to OrderService

File: src/main/java/online/bookstore/service/impl/OrderServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    private final OrderItemMapper orderItemMapper;
+    private final UserRepository userRepository;
+    private final ShoppingCartRepository shoppingCartRepository;
+
+    @Override
+    public Page<OrderDto> getAll(Long userId, Pageable pageable) {
+        return orderRepository.findAllByUserId(userId, pageable).map(orderMapper::toDto);
+    }
+
+    @Override
```
Comment (Elena-Bruyako) [inline src/main/java/online/bookstore/service/impl/OrderServiceImpl.java:44]: think where is Transactional needed

File: src/main/java/online/bookstore/service/impl/OrderServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    public OrderDto create(Long userId, String address) {
+        Order order = new Order();
+        order.setUser(userRepository.findById(userId)
+                .orElseThrow(() -> new EntityNotFoundException("Can't find user by id" + userId)));
+        order.setStatus(Order.Status.Pending);
+        order.setOrderItems(shoppingCartRepository.findById(userId).get().getCartItems().stream()
```
Comment (Elena-Bruyako) [inline src/main/java/online/bookstore/service/impl/OrderServiceImpl.java]: shop cart could be empty
improve readability

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see cooments

File: src/main/java/online/bookstore/service/impl/OrderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+    @Override
+    @Transactional
+    public OrderDto create(Long userId, CreateOrderRequestDto address) {
+        Order order = orderMapper.toModel(address);
+        order.setUser(userRepository.findById(userId)
+                .orElseThrow(() -> new EntityNotFoundException("Can't find user by id" + userId)));
+        order.setStatus(Order.Status.PENDING);
+        Set<CartItem> cartItems = shoppingCartRepository.findById(userId).get().getCartItems();
+        if (cartItems.isEmpty()) {
+            throw new RuntimeException("Shopping cart is empty: " + cartItems);
```
Comment (Elena-Bruyako) [inline src/main/java/online/bookstore/service/impl/OrderServiceImpl.java]:
```suggestion
            throw new OrderProcessingException("Shopping cart is empty for user " + userId);
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
one comment

**Review by ahoienko — APPROVED** [review]
Good job!

**Pull request merged**